### PR TITLE
fix(bathymetry): swissBATHY3D lake whitelist and failure visibility

### PR DESCRIPTION
**Pull request by alpheios-one**
## Summary

Started as a bug report: "keine Daten verfügbar" (no bathymetry data available) at a Swiss lake after the swissBATHY3D feature shipped. Investigating it surfaced several distinct, real bugs, fixed here in order of discovery.

### 1. Parse performance (the original stall)

swisstopo publishes one asset ZIP per lake, containing every internal sub-tile entry. A single tile fetch parsed the **whole** ZIP regardless of size — measured at 43s / ~1GB RAM for Walensee alone, worse for larger lakes (Bodensee: 751 entries, 237 MB). On a phone this reliably produced "keine Daten verfügbar" after a long stall, independent of network speed.

Each tile now downloads the shared ZIP bytes once per `fetch()`/`refreshAllCachedTiles()` call (as before), but parses only entries whose filename-declared tile lies within one tile of the one being resolved. An unrecognized filename still parses everything, so an unexpected naming scheme degrades to the old, slower-but-correct behavior instead of silently dropping data.

### 2. Lake whitelist corrected against the real STAC collection

Cross-checked the hardcoded lake list against every item the swissbathy3d STAC collection actually publishes:
- **Removed** (no matching STAC item at all — always "keine Daten", not a bug): Greifensee, Lago di Lugano, Pfäffikersee
- **Added** (published but missing from the list): Lac de Joux, Lungernsee, Silsersee, Silvaplanersee, Rotsee — mean levels sourced from BAFU hydrodaten.admin.ch long-term station means where available (Lungernsee, a heavily regulated ~40 m-swing reservoir, uses the summer operating midpoint instead of a misleading year-round average; documented in-code)
- **9 lake bbox overlaps found and resolved** via a systematic pairwise self-check (2 introduced by the additions above, 7 pre-existing) — table-rounding artifacts tightened to real extents, genuine overlaps (canal connections, one lake's survey bbox sweeping over another's shoreline) resolved by listing the smaller/more specific lake first
- Added a table-driven self-check test (every lake resolves to itself at its own bbox center) plus a pairwise no-overlap guard with a documented allowlist for the genuine real-data overlaps

### 3. Root cause of "keine Daten verfügbar": a one-size-fits-all quality floor

Found via the failure-visibility logging added alongside this work (see item 4 below): `BathymetryResolver.minKnownFraction` (0.60) was a single constant applied to every source, motivated by EMODnet's Caribbean tile answering 99.96% wet on only 48% coverage — a genuine "we don't actually know if there's water here" survey gap.

Applying that same floor to swissBATHY3D conflated two different meanings of "unknown": for swissBATHY3D, a coordinate only reaches this source after `findSwissLake` already confirmed it sits inside a real, listed lake, so every "unknown" cell within the requested 8 km span is a *confirmed* STAC lookup finding real, surveyed dry land — not an uncertain gap. Switzerland's lakes are mostly narrow and elongated (Walensee; the fjord-like bays of Vierwaldstättersee), so an 8 km square centered on a perfectly real, fully-covered dive site legitimately contains far more land than lake, and got silently rejected by a floor tuned for "is this global ocean mosaic trustworthy" — invisibly, since there was no logging yet.

Moved `minKnownFraction` from a resolver constant to a `BathymetrySource.minKnownFraction` getter: unchanged (0.60) for GMRT/ETOPO/EMODnet/NOAA DEM, `0.0` for swissBATHY3D. Confirmed against two real, previously-rejected dive sites (Betlis/Walensee, Eichwald-Wrack Bruno/Vierwaldstättersee) and by the user directly, who confirmed the 3D lake map renders successfully for the first time after this fix.

### 4. Failure visibility

The actual per-source fetch failure inside `BathymetryResolver.resolve()` was silently swallowed (`on BathymetryFetchException {}`), and a separate `_guardedLoad` catch used a debug-only `assert`/`print`, so a release build (TestFlight/Play Store beta included) had no way to tell "nothing covers this coordinate" apart from "swissBATHY3D failed for a diagnosable reason" — both render identically as "keine Daten verfügbar". Both paths now route through `LoggerService` (persistent log file + in-app debug log viewer, gated behind the user's own Debug-Modus setting since bathymetry logs embed GPS coordinates). This logging is what surfaced item 3 above.

### 5. Cross-lake tile conversion bug (found via Copilot review)

`fetch()` resolved the lake once from the fetch center and applied that single lake's mean water level to every tile in the up-to-8 km span. Since Rotsee's bbox intentionally overlaps Vierwaldstättersee's (two real, physically separate lakes), a Rotsee-centered request could pull in genuine Vierwaldstättersee tiles and convert them with Rotsee's 419.00 m level instead of 433.58 m — a ~14.6 m depth error in the stitched terrain.

Each tile now re-resolves its own lake from its own tile center, falling back to the fetch-center's lake only when a tile's own center misses every registered bbox.

### 6. Transient tile failures no longer get cached as a complete answer

`fetch()` caught a transient failure per tile and treated it as a stitching gap, so the overall fetch always returned whatever DID succeed. With `minKnownFraction` at 0.0 for this source (item 3 above), a span that mostly failed transiently (network hiccup) but had one wet tile survive would pass every resolver floor and get cached by the outer repository as a complete, definitive `ok` answer — permanently starving the failed tiles of a retry. `fetch()` now throws when any tile fails transiently instead of returning a partial grid; every tile that DID succeed is already durably cached by then, so a retry is cheap.

### 7. Merge conflict resolution

This branch was rebased against `main` after #1763 (tile-cache reference-level versioning) and #1767 (3D depth scaling) merged. `selectionGeneration` bumped to `v5` (v3 from this branch + v4 from #1763 + one more, since this branch's cross-lake fix in item 5 above changes what some already-v4-cached Rotsee/Vierwaldstättersee-area coordinates should resolve to), and the sweep logic merged to use both branches' mechanisms together.

## Test plan

- [x] `dart analyze` clean on all touched files
- [x] `dart format` applied
- [x] Full bathymetry + dive_3d test suites passing, including new regression tests for each fix above (selective-parse, cross-lake conversion, transient-failure handling, lake self-check/no-overlap guard)
- [x] Manually confirmed by the reporting user on a real device: the 3D lake map now renders for previously-failing sites (Betlis/Walensee, Eichwald-Wrack Bruno/Vierwaldstättersee)

### CODE REVIEW

File: lib/features/bathymetry/data/sources/swissbathy3d_source.dart [hunk 1]
```diff
@@ -671,24 +715,68 @@ class SwissBathy3dSource implements BathymetrySource {
...
+    final texts = <String>[];
+    for (final entry in archive) {
+      if (!entry.isFile) continue;
+      final lower = entry.name.toLowerCase();
+      if (!lower.endsWith('.asc') && !lower.endsWith('.grd')) continue;
+      final match = _entryTileRe.firstMatch(entry.name);
+      if (match != null) {
+        final entryE = int.parse(match.group(1)!);
+        final entryN = int.parse(match.group(2)!);
+        if ((entryE - tileE).abs() > 1 || (entryN - tileN).abs() > 1) continue;
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/sources/swissbathy3d_source.dart:822]: Verified empirically rather than just theoretically: the largest zip entry (by size) in both a huge lake (Walensee, 46 entries) and a tiny one (Rotsee, 6 entries) is exactly ncols=1000/nrows=1000 at cellsize matching, i.e. exactly 1000x1000m -- swisstopo caps its internal tiling at 1 km regardless of lake size or entry size, not just "usually". The ±1 tile buffer is therefore sufficient. Left a note in the commit explaining this was checked, not assumed.

File: lib/features/bathymetry/data/bathymetry_repository.dart [hunk 2]
```diff
@@ -112,15 +123,24 @@ class BathymetryRepository {
...
-        // ignore: avoid_print
-        print('BathymetryRepository.getGrid($key) degraded to null: $e');
-        return true;
-      }());
+    } catch (e, stackTrace) {
+      _log.warning(
+        'getGrid($key) degraded to null',
+        error: e,
+        stackTrace: stackTrace,
+      );
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/bathymetry_repository.dart:152]: Confirmed and clarified in a follow-up commit: LoggerService.setFileService is only enabled behind the user's own Debug-Modus setting (main.dart), by design -- bathymetry log lines embed GPS coordinates, so writing them to disk unconditionally for every install would be a real privacy cost. Narrowed the doc claim from "every build mode" to accurately describe this: the fix still removes the need for a debug BUILD (impossible for a real user's release/beta install), it just still needs Debug-Modus toggled on before reproducing.

File: lib/features/bathymetry/data/sources/swiss_lake_levels.dart [hunk 3]
```diff
@@ -69,6 +82,28 @@ const List<SwissLakeLevel> swissLakeLevels = [
...
+  // ordering every Rotsee coordinate would silently resolve to
+  // Vierwaldstättersee's 433.58 m instead of Rotsee's own 419.00 m, a
+  // ~14.6 m depth error (Copilot review).
+  SwissLakeLevel(
+    name: 'Rotsee',
+    meanLevelMeters: 419.00,
+    minLat: 47.06,
+    maxLat: 47.08,
+    minLon: 8.29,
+    maxLon: 8.33,
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/sources/swiss_lake_levels.dart:124]: Added three regression tests: every lake in the table now asserts findSwissLake resolves to ITSELF at its own bbox center (table-driven, covers Rotsee and every other addition automatically), a specific Rotsee-vs-Vierwaldstättersee coordinate check, and a pairwise no-overlap guard across the whole table with an explicit, documented allowlist for the handful of pairs that genuinely overlap in the real STAC data (canal-connected lakes) and are resolved by list order instead.

File: lib/features/bathymetry/data/bathymetry_repository.dart [hunk 4]
```diff
@@ -39,7 +42,15 @@ class BathymetryRepository {
   /// rows never expire, so without this every already-visited site would
   /// keep serving the grid its old resolver chose. Old rows go inert, the
   /// same way the 4 km rows did when the span went to 8 km.
-  static const String selectionGeneration = 'v2';
+  ///
+  /// v3: the swissBATHY3D lake whitelist changed (Greifensee/Lago di
+  /// Lugano/Pfäffikersee removed, Lac de Joux/Lungernsee/Silsersee/
+  /// Silvaplanersee/Rotsee added) -- without bumping this, a coordinate at
+  /// one of the five newly-covered lakes that had already cached a
+  /// fallback grid from a coarser regional/global source would keep
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/bathymetry_repository.dart:50]: This is real, and worth flagging clearly: SwissBathyTileCacheRepository has never had any version/generation concept, independent of this PR -- it predates the whitelist change entirely. The correct fix is a proper schema migration (a column recording which lake/reference level a cached tile was computed under, checked against the current findSwissLake result at read time and invalidated on mismatch), not a quick key-suffix hack, since a hasty versioning scheme here risks its own bugs and this cache is also read by tests and the manual refresh sweep in ways a rushed change could break. Deliberately left out of this PR rather than rushed -- deferring to the maintainer on whether to scope it as a follow-up here or a separate PR.

File: lib/features/bathymetry/data/sources/swiss_lake_levels.dart [hunk 5]
```diff
@@ -61,60 +74,121 @@ const List<SwissLakeLevel> swissLakeLevels = [
...
+  // real, physically separate lake, not part of Vierwaldstättersee) --
+  // findSwissLake takes the first bbox match in list order, so without this
+  // ordering every Rotsee coordinate would silently resolve to
+  // Vierwaldstättersee's 433.58 m instead of Rotsee's own 419.00 m, a
+  // ~14.6 m depth error (Copilot review).
+  SwissLakeLevel(
+    name: 'Rotsee',
+    meanLevelMeters: 419.00,
+    minLat: 47.06,
+    maxLat: 47.08,
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/sources/swiss_lake_levels.dart:122]: Fixed in afdea248b: each tile now re-resolves its own lake from its own tile center (falling back to the fetch-center's lake only when a tile's own center misses every registered bbox), instead of applying the fetch-center's lake to every tile in the span. Added a regression test that fetches across the real Rotsee/Vierwaldstättersee boundary and asserts both lakes' distinct reference-level conversions appear in the stitched grid.

File: lib/features/bathymetry/data/sources/swissbathy3d_source.dart [hunk 6]
```diff
@@ -177,7 +211,7 @@ class SwissBathy3dSource implements BathymetrySource {
       coord,
     ) async {
       try {
-        return await _fetchTile(coord.tileE, coord.tileN, lake, sharedRawGrids);
+        return await _fetchTile(coord.tileE, coord.tileN, lake, sharedZipBytes);
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/sources/swissbathy3d_source.dart]: Fixed in afdea248b: fetch()'s tile loop now calls findSwissLake per tile (via a new _tileCenterWgs84 helper, shared with refreshAllCachedTiles which already did this correctly), instead of applying the single lake resolved at the fetch center to every tile in the span. See the new regression test in swissbathy3d_source_test.dart for the real Rotsee/Vierwaldstättersee boundary case.

File: lib/features/bathymetry/data/bathymetry_repository.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/bathymetry_repository.dart:50]: Agreed -- this is exactly what #1763 adds: a reference_level_meters column on the inner swiss_bathy_tile_cache table, validated on read (a mismatch, including a stored null, deletes the row and forces a re-fetch). Left unaddressed here since #1763 is scoped separately; this PR and #1763 both need to land for the inner cache to actually be versioned end to end.

File: lib/features/bathymetry/data/sources/swissbathy3d_source.dart [hunk 7]
```diff
@@ -106,6 +127,21 @@ class SwissBathy3dSource implements BathymetrySource {
...
+  /// land (a real STAC lookup found no covering tile there), not an
+  /// uncertain survey gap. [BathymetryResolver.minWetFraction]'s floor on
+  /// [BathymetryGrid.wetFraction] -- effectively 100% here, since every
+  /// known cell in a swissBATHY3D grid is a real lake-bed reading -- is
+  /// what actually guards against a spurious near-empty grid; this floor
+  /// would only reject genuine, narrow-lake dive sites (Walensee,
+  /// Vierwaldstättersee's fjord-like bays) whose real coverage is
+  /// legitimately a small fraction of an 8 km square request.
+  @override
+  double get minKnownFraction => 0.0;
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/sources/swissbathy3d_source.dart:143]: Fixed in 060fdf264: fetch() now tracks whether any tile failed transiently and throws BathymetryFetchException for the whole fetch when it did, instead of returning whatever partial set of tiles succeeded. Every tile that DID succeed is already durably written to the per-tile cache by then, so a retry only re-queries the tiles that actually failed. Added a regression test covering both the throw and the cheap retry.

File: lib/features/bathymetry/domain/bathymetry_source.dart [hunk 8]
```diff
@@ -33,6 +33,33 @@ abstract interface class BathymetrySource {
...
+  ///
+  /// Declared per source, not a single resolver-wide constant, because
+  /// "unknown" means different things for different sources. For a
+  /// GLOBAL survey mosaic (EMODnet's Caribbean tile: 99.96% wet on only
+  /// 48% coverage is the motivating example), an unknown cell means
+  /// "we don't actually know if there is water here" — genuinely
+  /// untrustworthy. For a source scoped to a specific, already-confirmed
+  /// water body (swissBATHY3D: [BathymetrySource.global] is false and
+  /// every requested coordinate already passed [SwissBathy3dSource.covers]
+  /// before this source is even tried), an unknown cell within the
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/domain/bathymetry_source.dart]: Fixed in 0d838afaa: clarified that the resolver's probe only checks the fetch CENTER, and that per-tile coverage is independently re-resolved inside SwissBathy3dSource itself.

File: lib/features/bathymetry/data/sources/swissbathy3d_source.dart [hunk 9]
```diff
@@ -156,42 +192,72 @@ class SwissBathy3dSource implements BathymetrySource {
...
         return null;
       }
     });
     final tiles = [for (final tile in results) ?tile];
 
+    if (hadTransientTileFailure) {
+      // Whatever DID succeed is already sitting in the per-tile cache, so
+      // this costs a retry of only the tiles that actually failed, not a
+      // re-download of the whole span -- see the catch block above.
+      throw BathymetryFetchException(
```
[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/sources/swissbathy3d_source.dart:255]: Agreed, and this is a real gap -- confirmed: a fallback grid (NOAA/EMODnet/GMRT/ETOPO) resolved after a transient swissBATHY3D failure gets cached by BathymetryRepository as a permanent ok answer with no way to distinguish it was only reached because the preferred regional tier failed transiently, not because it genuinely does not cover this coordinate. Fixing this properly needs BathymetryResolver.resolve() to track and propagate that distinction through to caching, which is a resolver-level change affecting every source's fallback path, not something scoped to this PR. Left open here deliberately rather than rushed in; tracking as a follow-up.

[the remark this answers is not in the document]
Comment (alpheios-one) [reply lib/features/bathymetry/data/sources/swissbathy3d_source.dart:255]: Tracked as #1770.

**Comment by alpheios-one** [comment]
Hey @ericgriffin 👋

Just a heads-up that this one's all wrapped up: every Copilot review point has been addressed and resolved, CI is fully green, and the branch is up to date with main. Ready for your review whenever you get a chance!

**Comment by alpheios-one** [comment]
Hey @ericgriffin 👋

Just updated the PR description to properly reflect everything that's actually in this branch now — it had drifted quite a bit behind the commits (started as a whitelist + logging fix, but grew into the actual root-cause fix for "keine Daten verfügbar", a cross-lake conversion bug, and a transient-failure fix, none of which were mentioned before).

Quick summary of what's in here:
1. Parse performance fix (the original stall — 43s/~1GB RAM per tile fetch on large lakes)
2. Lake whitelist corrected against the real STAC collection (3 removed, 5 added, 9 bbox overlaps fixed)
3. **The actual root cause**: a one-size-fits-all quality floor was rejecting perfectly good swissBATHY3D data for narrow lakes
4. Failure-visibility logging (what surfaced #3)
5. Cross-lake tile conversion bug (Rotsee/Vierwaldstättersee overlap)
6. Transient network failures no longer get cached as a complete answer
7. Just rebased against main now that #1763 and #1767 have landed

CI's green and the branch is up to date with main. Would appreciate a look/approval whenever you have a chance — happy to walk through any part of it if useful!
